pkg/sentry/fsimpl/fuse/connection.go

```diff
@@ -65,9 +79,160 @@ type Response struct {
 type Connection struct {
 	fd *DeviceFD
 
-	// MaxWrite is the daemon's maximum size of a write buffer.
-	// This is negotiated during FUSE_INIT.
-	MaxWrite uint32
+	// initialized after receiving FUSE_INIT reply.
+	// Until it's set, suspend sending FUSE requests.
+	// Use SetInitialized() and IsInitialized() for atomic access.
+	initialized int32
+
+	// initializedChan is used to block requests before initialization.
+	initializedChan chan struct{}
+
+	// blocked when:
+	//   before the INIT reply is received (Initialized == false),
+	//   if there are too many outstading backgrounds requests (NumBackground == MaxBackground).
+	// TODO(gvisor.dev/issue/3185): use a channel to block.
+	blocked bool
+
+	// connected (connection established) when a new FUSE file system is created.
+	// Set to false when:
+	//   umount,
+	//   connection abort,
+	//   device release.
+	connected bool
+
+	// aborted via sysfs.
+	// TODO(gvisor.dev/issue/3185): abort all queued requests.
+	aborted bool
+
+	// connInitError if FUSE_INIT encountered error (major version mismatch).
+	// Only set in INIT.
+	connInitError bool
+
+	// connInitSuccess if FUSE_INIT is successful.
+	// Only set in INIT.
+	connInitSuccess bool
+
+	// TODO(gvisor.dev/issue/3185): All the queue logic are working in progress.
+
+	// NumberBackground is the number of requests in the background.
+	numBackground uint16
+
+	// congestionThreshold for NumBackground.
+	// Negotiated in FUSE_INIT.
+	congestionThreshold uint16
+
+	// maxBackground is the maximum number of NumBackground.
+	// Block connection when it is reached.
+	// Negotiated in FUSE_INIT.
+	maxBackground uint16
+
+	// numActiveBackground is the number of requests in background and has being marked as active.
+	numActiveBackground uint16
+
+	// numWating is the number of requests waiting for completion.
+	numWaiting uint32
+
+	// TODO(gvisor.dev/issue/3185): BgQueue
+	// some queue for background queued requests.
+
+	// bgLock protects:
+	// MaxBackground, CongestionThreshold, NumBackground,
+	// NumActiveBackground, BgQueue, Blocked.
+	bgLock sync.Mutex
+
+	// maxRead is the maximum size of a read buffer in in bytes.
+	maxRead uint32
+
+	// maxWrite is the maximum size of a write buffer in bytes.
+	// Negotiated in FUSE_INIT.
+	maxWrite uint32
+
+	// maxPages is the maximum number of pages for a single request to use.
+	// Negotiated in FUSE_INIT.
+	maxPages uint16
+
+	// minor version of the FUSE protocol.
+	// Negotiated and only set in INIT.
+	minor uint32
+
+	// asyncRead if read pages asynchronously.
+	// Negotiated and only set in INIT.
+	asyncRead bool
+
+	// abortErr is true if kernel need to return an unique read error after abort.
+	// Negotiated and only set in INIT.
+	abortErr bool
+
+	// atomicOTrunc is true when FUSE does not send a separate SETATTR request
+	// before open with O_TRUNC flag.
+	// Negotiated and only set in INIT.
+	atomicOTrunc bool
```

Comment:
I'm a bit hesitant to define all these feature flags in this patch, when we don't actually implement any of them.

On the one hand, FUSE_INIT is where these are negotiated, so one could argue it's logical to remember the negotiated values. But on the other hand, no one uses these values yet, and having them defined like this gives a reader the impression that these are all fully implemented.

I think we should remove most unsupported flags from this patch, and add them in as we implement them. I think the following flags should not be included until we implement them, as they involve pretty specific bits of logic not relevant to overall operation of a FUSE filesystem:

- atomicOTrunc (add once we support open(O_TRUNC))
- exportSupport
- parallelDirOps
- handleKillPriv
- noLock (add when we implement POSIX locks)
- noFlocks (add when we implement POSIX locks)
- autoInvalData (add when we implement page cache eviction for FUSE)
- explicitInvalidData (I don't really understand what this flag implies, but we certainly don't handle it right now)
- doReaddirPlus / readdirPlusAuto (add when we implement FUSE_READDIRPLUS)
- asyncDIO (we are unlikely to ever support this)
- posixACL

The rest seem general enough that we can keep here.

Feel free to add the list of currently unsupported flags as a comment here, so we can easily reference it as we implement them:

```
// The following FUSE_INIT flags are currently unsupported by this implementation:
- FUSE_ATOMIC_O_TRUNC: requires open(..., O_TRUNC)
- FUSE_PARALLEL_DIROPS
...
```